### PR TITLE
feat(health): support various policies _.microgateway.airlock.com

### PR DESCRIPTION
**Pull request by root30**
This PR adds the custom health check for various policy Types of Airlock Microgateway.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: resource_customizations/microgateway.airlock.com/_/testdata/xcustomresource_unknown.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+apiVersion: microgateway.airlock.com/v1alpha1
+kind: XCustomResource
```
Comment (crenshaw-dev) [inline resource_customizations/microgateway.airlock.com/_/testdata/xcustomresource_unknown.yaml]: I'm not loving the idea of using contrived resource kinds for tests. The wildcard pattern allows us to avoid duplicating a lot of script logic. But it shouldn't be used as a way to avoid writing thorough tests. 

Is there a strong reason why we can't use actual examples for the tests?

File: resource_customizations/microgateway.airlock.com/_/health.lua [hunk 2]
```diff
@@ -0,0 +1,49 @@
+local hs = { status = "Progressing", message = "Waiting for " .. (obj.kind or "Policy")  .. " status"}
+
+local function is_policy_kind(kind)
+  return kind ~= nil and string.match(kind, "Policy$") ~= nil
```
Comment (crenshaw-dev) [inline resource_customizations/microgateway.airlock.com/_/health.lua:4]: I don't like the idea of matching a kind suffix. If a new kind is added that matches this suffix, there's no reason to believe it'll necessarily follow the same status field conventions. I'd recommend instead hard-coding a static list of kinds that we know this check supports.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: resource_customizations/microgateway.airlock.com/_/health.lua (hunk 2 above)
Comment (root30) [reply resource_customizations/microgateway.airlock.com/_/health.lua:4]: Hi @crenshaw-dev 

Thanks for raising these concerns.

For the first point, the decision to avoid using a fixed list of concrete resource kinds was intentional. We want to avoid having to update the HealthCheck implementation every time a new policy kind is introduced, as that would require a separate PR just to register the new kind. The wildcard-based approach allows the HealthCheck to automatically support additional policy resources as they are added to our implementation.

Regarding the tests, the goal was not to avoid thorough test coverage. The additional policies we plan to support are Gateway API direct attached policies, which follow the same status conventions as existing resources such as [BackendTLSPolicy](https://gateway-api.sigs.k8s.io/reference/api-types/policy/backendtlspolicy/). Because the HealthCheck logic is based on those shared conventions rather than kind-specific behavior, we considered it more valuable to test the common behavior than to duplicate equivalent tests for every individual policy kind.

For the second point, we agree that blindly matching arbitrary suffixes would be risky if unrelated resources could satisfy the pattern. In this case, however, the expectation is that future matching resources will be Gateway API policy kinds that share the same status structure and semantics. The intent behind the suffix matching is to make the HealthCheck extensible without requiring code changes whenever a new policy kind is introduced. If we hard-code the supported kinds, we lose that extensibility and would need to keep the list synchronized as our implementation evolves.

File: resource_customizations/microgateway.airlock.com/_/testdata/xcustomresource_unknown.yaml (hunk 1 above)
Comment (root30) [reply resource_customizations/microgateway.airlock.com/_/testdata/xcustomresource_unknown.yaml]: Please see my detailed response in the other conversation.

However regarding your question: We can use an actual custom resource of ours for the `unknown` test case if this is the sole concern.

File: resource_customizations/microgateway.airlock.com/_/health.lua (hunk 2 above)
Comment (crenshaw-dev) [reply resource_customizations/microgateway.airlock.com/_/health.lua:4]: Are you a maintainer of microgateway? I want to make sure we have solid reason to believe the future CRDs will follow existing conventions. A link to a docs page stating this would be extra valuable.  :-)

Comment (root30) [reply resource_customizations/microgateway.airlock.com/_/health.lua:4]: Yes, I am one of the maintainers of Microgateway.

The standardized Policy resource of GatewayAPI implemented in Microgateway is specified by [GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/). For the detailed definition of the status field conventions please take a look at the [Status reporting](https://gateway-api.sigs.k8s.io/geps/gep-713/#status-reporting) section.

We really appreciate your due diligence and are also interested in a robust and low-maintenance solution that benefits everyone.
If you have any further questions or concerns, we are open to discuss them.

Comment (crenshaw-dev) [reply resource_customizations/microgateway.airlock.com/_/health.lua:4]: Perfect, thanks for confirming!

File: resource_customizations/microgateway.airlock.com/_/health.lua [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  for i, ancestor in ipairs(obj.status.ancestors) do
+    for j, condition in ipairs(ancestor.conditions) do
+      if condition.type == "Accepted" then
+        if condition.status ~= "True" then
+          hs.status = "Degraded"
+          hs.message = "Ancestor " .. (ancestor.ancestorRef.name or "") .. ": " .. condition.message
+          return hs
+        else
+          hs.status = "Healthy"
+          hs.message = (obj.kind or "Policy")  .. " is healthy"
```
Comment (crenshaw-dev) [inline resource_customizations/microgateway.airlock.com/_/health.lua:34]: My reading of this is that if any ancestor is healthy, we'll basically ignore any ancestors that are still progressing but that do not yet have the Accepted condition yet... Maybe the "Healthy" path should be a final loop that confirms that _all_ ancestors are healthy?

Comment (root30) [reply resource_customizations/microgateway.airlock.com/_/health.lua:34]: You are right! I added a test case for that behavior and changed the logic accordingly.

File: resource_customizations/microgateway.airlock.com/_/testdata/unknown.yaml [hunk 4]
```diff
@@ -14,5 +14,4 @@ status:
       group: gateway.networking.k8s.io
       kind: Gateway
       name: example-gateway
-    conditions: []
```
Comment (root30) [inline resource_customizations/microgateway.airlock.com/_/testdata/unknown.yaml]: An empty conditions struct is not valid (according to [policy_types.go](https://github.com/kubernetes-sigs/gateway-api/blob/f0cff36a55397106c2c3052f64adca53a3526908/apis/v1/policy_types.go#L240)), however `conditions` would be required within the `ancestors` struct as well... in that case it does not matter though

File: resource_customizations/microgateway.airlock.com/_/health.lua (hunk 3 above)
Comment (crenshaw-dev) [reply resource_customizations/microgateway.airlock.com/_/health.lua:34]: I think we still have an ordering issue here. `not hasAcceptedCondition` covers the case where no ancestors have an accepted condition, but it doesn't cover the case where some do have an accepted condition and others do not. Presumably "no accepted condition" means "Progressing." But the way this is currently written, if one out of 10 ancestors have a healthy accepted condition, then the whole resource is Healthy.

Comment (root30) [reply resource_customizations/microgateway.airlock.com/_/health.lua:34]: hmm, I am not sure I really got your point.

I believe we loop over all ancestors and resetting the `hasAcceptedCondition` for every new iteration (line 23/24). So `not hasAcceptedCondition` should yield `true` when one out of all ancestors does not have an accepted condition.
If one ancestor has an accepted condition (else clause on line 32) we do not return but only set `status` and `message` fields to return in the end once all checks passed successful.

The test case [degraded_multiple_ancestors](https://github.com/root30/argo-cd/blob/8c28c0891c9525a387dc0683e90e63ddae09b90f/resource_customizations/microgateway.airlock.com/_/testdata/degraded_multiple_ancestors.yaml) should exatly cover the discussed scenario:
Multiple ancestors, one of them having `Accepted=True` and another without an `Accepted` Condition
```
...
    conditions:
    - lastTransitionTime: '2026-04-16T12:20:11Z'
      message: RateLimitPolicy is accepted
      observedGeneration: 3
      reason: Accepted
      status: 'True'
      type: Accepted
    controllerName: example.com/gatewayclass-controller
  - ancestorRef:
      group: gateway.networking.k8s.io
      kind: Gateway
      name: example-gateway-2
    conditions:
      - lastTransitionTime: '2026-04-16T12:20:11Z'
        message: FooBar
        observedGeneration: 3
        reason: Foo
        status: 'True'
        type: Bar
    controllerName: example.com/gatewayclass-controller
```

Comment (crenshaw-dev) [reply resource_customizations/microgateway.airlock.com/_/health.lua:34]: Ah! I misread the nested loop. Thanks for clarifying!

**Review by crenshaw-dev — APPROVED** [review]
